app/src/lib/lookups.js

```diff
@@ -1,37 +1,152 @@
-import {TOKEN_TYPE, flatten_sentence} from './parser/token'
+import {TOKEN_TYPE} from './parser/token'
+
+// TODO move whole token pipeline to the server
+
+/**
+ * 
+ * @param {Sentence[]} sentences 
+ * @returns {Promise<Sentence[]>}
+ */
+export async function perform_form_lookups(sentences) {
+	const lookup_tokens = sentences.flatMap(flatten_for_lookup).filter(is_lookup_token)
+
+	await Promise.all(lookup_tokens.map(check_forms))
+
+	return sentences
+}
 
 /**
  * 
  * @param {Sentence[]} sentences 
  * @returns {Promise<Sentence[]>}
  */
 export async function perform_ontology_lookups(sentences) {
-	const lookup_tokens = sentences.flatMap(flatten_sentence).filter(is_lookup_token)
+	const lookup_tokens = sentences.flatMap(flatten_for_lookup).filter(is_lookup_token)
 
 	await Promise.all(lookup_tokens.map(check_ontology))
 
 	return sentences
+}
+
+/**
+ * 
+ * @param {Sentence} sentence 
+ * @returns {Token[]}
+ */
+function flatten_for_lookup(sentence) {
```

Comment:
This fixes the pairing issue from #40. I figured it was simple enough to fix it here, but let me know if you'd rather it be separate.